MudTabs: fix index out error and improve the use experience

ssccinng opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->


## Description
<!-- Describe your changes in detail any why -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #1337 or fixes #9310 -->

## How Has This Been Tested?
<!-- All PR's should implement unit tests if possible -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

<!-- If you made any visual changes, provide screenshots of before/after, it its moving parts, please provide high quality gif, wemb or mp4 -->

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [ ] I've added relevant tests.

ssccinng committed:
[Fix] Prevent cross-border error reporting and improve the use experience

ssccinng committed:
Delete Debug

JonBunator reviewed (changes requested):
I am not too sure what this PR is about? Can you please explain further?
Can you also add a testcase please?

ssccinng commented:
> I am not too sure what this PR is about? Can you please explain further? Can you also add a testcase please?

When I use mudtabs, if I use code similar to the following and update source
```
<MudTabs Elevation="0" ActivePanelIndex=@(pclRound.Swissidx - 1)>

    @for (int swidx = 0; swidx < pclRound.Swissidx; ++swidx)
   {
        int sswidx = swidx;
        ....
   }
</MudTabs>
```
ActivePanelIndex appears to become out-of-range values before the panels are updated, causing an error to be reported. So I added this modification so that ActivePanelIndex doesn't report an error when it's out of range and automatically switches to the correct page when the panels are updated

JonBunator commented:
> > I am not too sure what this PR is about? Can you please explain further? Can you also add a testcase please?
> 
> When I use mudtabs, if I use code similar to the following and update source（pclRound）
> ```
> <MudTabs Elevation="0" ActivePanelIndex=@(pclRound.Swissidx - 1)>
> 
>     @for (int swidx = 0; swidx < pclRound.Swissidx; ++swidx)
>    {
>         int sswidx = swidx;
>         ....
>     <MudTabPanel>
>         ....
>     </MudTabPanel >
>    }
> </MudTabs>
> ```
> ActivePanelIndex appears to become out-of-range values before the panels are updated, causing an error to be reported. So I added this modification so that ActivePanelIndex doesn't report an error when it's out of range and automatically switches to the correct page when the panels are updated
> 

Can you please provide a try.mudblazor.com reproduction and a description of your use case?

ssccinng commented:
[> > > I am not too sure what this PR is about? Can you please explain further? Can you also add a testcase please?
> > 
> > 
> > When I use mudtabs, if I use code similar to the following and update source（pclRound）
> > ```
> > <MudTabs Elevation="0" ActivePanelIndex=@(pclRound.Swissidx - 1)>
> > 
> >     @for (int swidx = 0; swidx < pclRound.Swissidx; ++swidx)
> >    {
> >         int sswidx = swidx;
> >         ....
> >     <MudTabPanel>
> >         ....
> >     </MudTabPanel >
> >    }
> > </MudTabs>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ActivePanelIndex appears to become out-of-range values before the panels are updated, causing an error to be reported. So I added this modification so that ActivePanelIndex doesn't report an error when it's out of range and automatically switches to the correct page when the panels are updated
> 
> Can you please provide a try.mudblazor.com reproduction and a description of your use case?

https://try.mudblazor.com/snippet/GuQmYhQupLYIRMrL](https://try.mudblazor.com/snippet/GuQmYhQupLYIRMrL)

> > > I am not too sure what this PR is about? Can you please explain further? Can you also add a testcase please?
> > 
> > 
> > When I use mudtabs, if I use code similar to the following and update source（pclRound）
> > ```
> > <MudTabs Elevation="0" ActivePanelIndex=@(pclRound.Swissidx - 1)>
> > 
> >     @for (int swidx = 0; swidx < pclRound.Swissidx; ++swidx)
> >    {
> >         int sswidx = swidx;
> >         ....
> >     <MudTabPanel>
> >         ....
> >     </MudTabPanel >
> >    }
> > </MudTabs>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ActivePanelIndex appears to become out-of-range values before the panels are updated, causing an error to be reported. So I added this modification so that ActivePanelIndex doesn't report an error when it's out of range and automatically switches to the correct page when the panels are updated
> 
> Can you please provide a try.mudblazor.com reproduction and a description of your use case?

https://try.mudblazor.com/snippet/GuQmYhQupLYIRMrL

JonBunator commented:
Tabs are not meant to be used with a for loop. Use MudDynamicTabs for it instead

ssccinng commented:
> Tabs are not meant to be used with a for loop. Use MudDynamicTabs for it instead

I just read the official website example, but I think this operation seems to be a little troublesome. If I deal with tabs a little, the code can be well compatible with this operation

ssccinng commented:
> Tabs are not meant to be used with a for loop. Use MudDynamicTabs for it instead
Sorry, I don't have much experience in PR and don't know how to provide test cases for this. In my personal project, I think this function is very useful. Can muddynamictabs achieve a similar effect? I think the example on the official website is also difficult to achieve this.

ssccinng commented:
> Tabs are not meant to be used with a for loop. Use MudDynamicTabs for it instead

Sorry, I don't have much experience in PR and don't know how to provide test cases for this. In my personal project, I think this function is very useful. Can muddynamictabs achieve a similar effect? I think the example on the official website is also difficult to achieve this.

just-the-benno commented:
My two cents

I think the developer should be responsible for setting the correct index: an index within the boundary of the tab control. (Which includes "normal" and dynamic tabs).  

@ssccinng proposal only includes the upper bound (> Count) but not the lower bound (< 0).  Besides, the error handling, by not displaying a tab at all, might help prevent throwing an exception and potential crash of the app or session but also "hides" the underlying exception. I think this exception is important to present because something severe must have happened to create this situation in the first place.

``MudTabs`` already exposes the items as an ``IEnumerable``. So, the developer has already a powerful means to check the boundaries before setting the selected index. This behavior can be reused by creating an extension method.

While I understand the presentated scenario, I think the library has already all tools needed to prevent it from happing and hence we should close this PR.

ssccinng commented:
> 

<0 is not a normal index, so I didn't modify it.

When I used radzen Blazer before, I wouldn't have similar problems. At present, I don't know how to modify my code to adapt to this situation easily. This problem seems to be caused by the refresh sequence. I don't know how to deal with this problem gracefully.

ssccinng commented:
> <0 is not a normal index, so I didn't modify it.
> 
> When I used Radzen Blazer before, I wouldn't have similar problems. At present, I don't know how to modify my code to adapt to this situation easily. This problem seems to be caused by the refresh sequence. I don't know how to deal with this problem gracefully.

<0 is not a normal index, so I didn't modify it.

When I used Radzen Blazer before, I wouldn't have similar problems. At present, I don't know how to modify my code to adapt to this situation easily. This problem seems to be caused by the refresh sequence. I don't know how to deal with this problem gracefully.

ssccinng commented:
> My two cents
> 
> I think the developer should be responsible for setting the correct index: an index within the boundary of the tab control. (Which includes "normal" and dynamic tabs).
> 
> @ssccinng proposal only includes the upper bound (> Count) but not the lower bound (< 0). Besides, the error handling, by not displaying a tab at all, might help prevent throwing an exception and potential crash of the app or session but also "hides" the underlying exception. I think this exception is important to present because something severe must have happened to create this situation in the first place.
> 
> `MudTabs` already exposes the items as an `IEnumerable`. So, the developer has already a powerful means to check the boundaries before setting the selected index. This behavior can be reused by creating an extension method.
> 
> While I understand the presentated scenario, I think the library has already all tools needed to prevent it from happing and hence we should close this PR.

<0 is not a normal index, so I didn't modify it.

When I used Radzen Blazer before, I wouldn't have similar problems. At present, I don't know how to modify my code to adapt to this situation easily. This problem seems to be caused by the refresh sequence. I don't know how to deal with this problem gracefully.

just-the-benno commented:
Have a look at this snippet here: 

https://try.mudblazor.com/snippet/muQmaWkHFjUmGqCR

If you want to select a newly added tab (I assume that is your problem), you need to make an extra step in your scenario.

If create your tabs using a ``for`` (or ``foreach``) loop. This new tab won't be part of the tab control immediately. (Hence the exception). You need to defer the new tab selection until it is ready. So, when you add the new tab, you save the expected index, and as soon as ``OnAfterRender`` is called, you set the new index and start the render cycle again.

In case I missed your scenario, I'd be great if you could provide a trymud, so we can work with that.

ssccinng commented:
> 



> Have a look at this snippet here:
> 
> https://try.mudblazor.com/snippet/muQmaWkHFjUmGqCR
> 
> If you want to select a newly added tab (I assume that is your problem), you need to make an extra step in your scenario.
> 
> If create your tabs using a `for` (or `foreach`) loop. This new tab won't be part of the tab control immediately. (Hence the exception). You need to defer the new tab selection until it is ready. So, when you add the new tab, you save the expected index, and as soon as `OnAfterRender` is called, you set the new index and start the render cycle again.
> 
> In case I missed your scenario, I'd be great if you could provide a trymud, so we can work with that.

Thanks. I know your processing method now, but if I use this method in my project, I think it will become very troublesome. I will process a lot of these indexes in OnAfterRender. If the processing method in PR is not good, can I add a property to adapt this matter? I really think it will be much more convenient to deal with `for` or `foreach` like this.

https://github.com/ssccinng/PokeSci/blob/500b12850b6297e09b3378db6099c22663ddf520/PokemonIsshoni.Net/Client/Pages/MatchPage/MatchManager.razor#L271
and
https://github.com/ssccinng/PokeSci/blob/master/PokemonIsshoni.Net/Client/Pages/MatchPage/MatchManager.razor.cs#L293

just-the-benno commented:
If you have an idea how you can do it elegantly, go ahead. 

I still think tabs are not supposed to be used that way. One way is to go ahead with DynamicTabs. 

There is very much in your component regarding your source code. Maybe it is a good idea to split it into smaller components, You can wrap the logic from the trymud into your own component or do stuff with inheritance to make it a bit smarter.

ssccinng commented:
> If you have an idea how you can do it elegantly, go ahead.
> 
> I still think tabs are not supposed to be used that way. One way is to go ahead with DynamicTabs.
> 
> There is very much in your component regarding your source code. Maybe it is a good idea to split it into smaller components, You can wrap the logic from the trymud into your own component or do stuff with inheritance to make it a bit smarter.

Indeed, my code is a little bloated, and I'm still exploring a better model,

Just now I went to see Radzen's processing method, which also adopts a similar processing method. Personally, I think it's good to support this grammar, but there may be a better way to realize it. I may need to explore.

just-the-benno commented:
Let me know if you are ready for another review :)

mikes-gh commented:
@ssccinng I see there are other ways to resolve this without  adding cyclometric complexity to already complex component.
Therefore I am going to close.  If you wish to come back to @just-the-benno we can further review but I am going to close for now.
